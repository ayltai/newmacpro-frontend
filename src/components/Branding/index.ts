export { default, } from './Branding';
export type { BrandingProps, } from './Branding.types';

