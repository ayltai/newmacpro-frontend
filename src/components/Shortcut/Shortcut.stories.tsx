import type { Meta, StoryObj, } from '@storybook/react';

import Shortcut from './Shortcut';

const meta : Meta<typeof Shortcut> = {
    component : Shortcut,
    tags      : [
        'autodocs',
    ],
    argTypes  : {
        mode     : {
            description : 'Specify the display mode to use. `desktop` is a text button. `mobile` is a menu item. If not provided, the mode is determined by the screen size.',
        },
        anchor   : {
            description : 'The `href` attribute of the anchored element to navigate to if `onClick` is not provided.',
        },
        onClick  : {
            description : 'An event handler called when the component is clicked.',
        },
        children : {
            control     : false,
            description : 'The content of the component.',
        },
    },
};

export default meta;

type Story = StoryObj<typeof Shortcut>;

/**
 * Desktop mode.
 */
export const Desktop : Story = {
    args : {
        mode     : 'desktop',
        children : 'Desktop mode',
    },
};

/**
 * Mobile mode.
 */
export const Mobile : Story = {
    args : {
        mode     : 'mobile',
        children : 'Mobile mode',
    },
};
