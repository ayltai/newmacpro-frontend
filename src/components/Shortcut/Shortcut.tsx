import { Button, MenuItem, useMediaQuery, } from '@mui/material';
import React, { FC, } from 'react';

import type { ShortcutProps, } from './Shortcut.types';

/**
 * Shortcuts allow users to quickly navigate to anchored elements.
 * @param mode Specify the display mode to use. `desktop` is a text button. `mobile` is a menu item. If not provided, the mode is determined by the screen size.
 * @param anchor The `href` attribute of the anchored element to navigate to if `onClick` is not provided.
 * @param onClick The event handler called when the component is clicked.
 * @param children The content of the component.
 * @param rest The other props passed down to the component.
 */
const Shortcut : FC<ShortcutProps> = ({
    mode,
    anchor,
    children,
    onClick,
    ...rest
}) => {
    const isDesktop   = useMediaQuery('(min-width: 900px)');
    const desktopMode = mode === 'desktop' || mode !== 'mobile' && isDesktop;

    const handleClick = () => window.location.href = `#${anchor ?? ''}`;

    return desktopMode ? (
        <Button
            size='large'
            color='inherit'
            onClick={onClick ?? handleClick}
            {...rest}>
            {children}
        </Button>
    ) : (
        <MenuItem
            onClick={onClick ?? handleClick}
            {...rest}>
            {children}
        </MenuItem>
    );
};

export default Shortcut;
