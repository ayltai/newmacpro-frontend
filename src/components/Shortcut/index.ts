export { default, } from './Shortcut';
export type { ShortcutProps, } from './Shortcut.types';
