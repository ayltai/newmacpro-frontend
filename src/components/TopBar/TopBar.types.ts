import { ReactElement, } from 'react';

import type { ActionProps, } from '../Action';
import type { BrandingProps, } from '../Branding';
import type { ShortcutProps, } from '../Shortcut';

export interface TopBarProps {
    maxWidth? : 'xs' | 'sm' | 'md' | 'lg' | 'xl',
    children? : ReactElement<ActionProps | BrandingProps | ShortcutProps> | ReactElement<ActionProps | BrandingProps | ShortcutProps>[],
}
