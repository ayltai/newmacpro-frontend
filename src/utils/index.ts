import * as Sentry from '@sentry/react';

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const capitalizeFirstLetter = (input : string) => input.charAt(0).toUpperCase() + input.slice(1);
