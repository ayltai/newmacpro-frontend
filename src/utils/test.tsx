import { ThemeProvider, } from '@mui/material';
import { configureStore, Store, } from '@reduxjs/toolkit';
import { act, render, } from '@testing-library/react';
import mediaQuery from 'css-mediaquery';
import { ConfirmProvider, } from 'material-ui-confirm';
import React, { ReactNode, } from 'react';
import { Provider, } from 'react-redux';

import { bundleApi, cartApi, } from '../apis';
import { POPULAR_APPS, } from '../constants';
import { cartReducer, prefReducer, store as reduxStore, } from '../states';
import { AppTheme, } from '../styles';
import { Cart, } from '../types';

export const mockApis = async (cart? : Cart) => {
    fetchMock.doMock(request => {
        if (request.url.endsWith('/api/bundles') && request.method === 'POST') return Promise.resolve({
            status : 201,
            body   : 'Created',
        });

        if (request.url.endsWith('/api/bundles/1') && request.method === 'GET') return Promise.resolve({
            status : 200,
            body   : JSON.stringify({
                id          : '1',
                displayName : 'Dummy',
                packages    : [
                    POPULAR_APPS[0],
                ],
            }),
        });

        if (request.url.endsWith('/api/bundles/1') && request.method === 'DELETE') return Promise.resolve({
            status : 200,
            body   : 'OK',
        });

        if (request.url.endsWith('/api/carts') && request.method === 'POST') return Promise.resolve({
            status : 201,
            body   : 'Created',
        });

        return Promise.resolve({
            status : 404,
            body   : 'Not found',
        });
    });

    await act(async () => {
        await reduxStore.dispatch(bundleApi.endpoints?.getBundle.initiate({
            bundleId : '1',
            idToken  : 'idToken',
        }));
        if (cart) await reduxStore.dispatch(cartApi.endpoints.addCart.initiate(cart));
    });
};

export const createMatchMedia = (width : number) => (query : string) => ({
    matches             : mediaQuery.match(query, {
        width,
    }),
    media               : '',
    onchange            : null,
    addListener         : () => {},
    removeListener      : () => {},
    addEventListener    : () => {},
    removeEventListener : () => {},
    dispatchEvent       : () => false,
});

export const createStore = (preloadedState? : Record<string, any>) => configureStore({
    preloadedState,
    reducer    : {
        cart                      : cartReducer,
        pref                      : prefReducer,
        [ bundleApi.reducerPath ] : bundleApi.reducer,
        [ cartApi.reducerPath   ] : cartApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(bundleApi.middleware, cartApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <ThemeProvider theme={AppTheme}>
                <ConfirmProvider>
                    {children}
                </ConfirmProvider>
            </ThemeProvider>
        </Provider>
    ),
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
