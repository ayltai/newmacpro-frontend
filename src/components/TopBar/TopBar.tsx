import { AppBar, Box, Button, Container, Toolbar, useMediaQuery, useScrollTrigger, } from '@mui/material';
import React, { cloneElement, FC, Fragment, ReactElement, } from 'react';

import { default as Action, ActionProps, } from '../Action';
import { default as Branding, BrandingProps, } from '../Branding';
import { default as Shortcut, ShortcutProps, } from '../Shortcut';
import { default as ShortcutMenu, } from '../ShortcutMenu';
import type { TopBarProps, } from './TopBar.types';

const ElevationScroll = ({
    children,
} : {
    children : ReactElement,
}) => {
    const trigger = useScrollTrigger({
        disableHysteresis : true,
        threshold         : 0,
    });

    return cloneElement(children, {
        elevation : trigger ? 4 : 0,
    });
};

const TopBar : FC<TopBarProps> = ({
    maxWidth = 'lg',
    children,
}) => {
    const desktopMode = useMediaQuery('(min-width:900px)');

    let branding : ReactElement<BrandingProps> | undefined;

    const actions   : ReactElement<ActionProps>[]   = [];
    const shortcuts : ReactElement<ShortcutProps>[] = [];

    if (children) {
        if (Array.isArray(children)) {
            branding = children.find(child => child.type === Branding) as ReactElement<BrandingProps>;

            actions.push(...children.filter(child => child.type === Action) as ReactElement<ActionProps>[]);
            shortcuts.push(...children.filter(child => child.type === Shortcut) as ReactElement<ShortcutProps>[]);
        } else {
            if (children.type === Branding) branding = children as ReactElement<BrandingProps>;

            if (children.type === Action) actions.push(children as ReactElement<ActionProps>);
            if (children.type === Shortcut) shortcuts.push(children as ReactElement<ShortcutProps>);
        }
    }

    const handleClick = () => window.location.href = '/';

    return (
        <Fragment>
            <ElevationScroll>
                <AppBar color='inherit'>
                    <Container maxWidth={maxWidth}>
                        <Toolbar disableGutters>
                            <Box flexGrow={1}>
                                {!desktopMode && shortcuts.length > 0 && (
                                    <ShortcutMenu>
                                        {shortcuts}
                                    </ShortcutMenu>
                                )}
                                {branding && (
                                    <Button
                                        size='large'
                                        color='inherit'
                                        onClick={handleClick}>
                                        {branding}
                                    </Button>
                                )}
                            </Box>
                            <Box flexGrow={0}>
                                {desktopMode && shortcuts}
                                {actions}
                            </Box>
                        </Toolbar>
                    </Container>
                </AppBar>
            </ElevationScroll>
            <Toolbar />
        </Fragment>
    );
};

export default TopBar;
