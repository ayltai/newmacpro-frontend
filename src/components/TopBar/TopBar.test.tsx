import React from 'react';

import { render, } from '../../utils/test';
import { default as Branding, } from '../Branding';
import { default as Shortcut, } from '../Shortcut';
import TopBar from './TopBar';

describe('<TopBar />', () => {
    it('renders correctly', () => {
        expect(render(
            <TopBar>
                <Branding>Branding</Branding>
                <Shortcut anchor='shortcut-1'>Shortcut 1</Shortcut>
                <Shortcut anchor='shortcut-2'>Shortcut 2</Shortcut>
            </TopBar>
        )).toMatchSnapshot();
    });
});
