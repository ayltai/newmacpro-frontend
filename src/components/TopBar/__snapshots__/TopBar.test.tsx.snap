// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopBar /> renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorInherit MuiAppBar-positionFixed mui-fixed css-yiikg7-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular css-r6ewbb-MuiToolbar-root"
          >
            <div
              class="MuiBox-root css-i9gxme"
            >
              <button
                aria-label="menu"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-184mzhy-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="MenuIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-colorInherit MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-colorInherit css-1g2vlp7-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Branding
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <div
              class="MuiBox-root css-2uchni"
            />
          </div>
        </div>
      </header>
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      />
    </div>
  </body>,
  "container": <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorInherit MuiAppBar-positionFixed mui-fixed css-yiikg7-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular css-r6ewbb-MuiToolbar-root"
        >
          <div
            class="MuiBox-root css-i9gxme"
          >
            <button
              aria-label="menu"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-184mzhy-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-colorInherit MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-colorInherit css-1g2vlp7-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Branding
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiBox-root css-2uchni"
          />
        </div>
      </div>
    </header>
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
